pkg/collector/corechecks/system/wincrashdetect/probe/crashparse.go

```diff
@@ -0,0 +1,135 @@
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2021-present Datadog, Inc.
+//go:build windows
+// +build windows
+
+package probe
+
+/*
+#cgo LDFLAGS: -l dbgeng -static
+#include "crashdump.h"
+*/
+import "C"
+import (
+	"fmt"
+	"strings"
+	"unsafe"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+type logCallbackContext struct {
+	loglines       []string
+	hasSeenRetAddr bool
+	unfinished     string
+}
+
+const maxscan = int(200)
+
+func min(a, b int) int {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+const (
+	bugcheckCodePrefix     = "Bugcheck code"
+	debugSessionTimePrefix = "Debug session time"
+)
+
+//export logLineCallback
+func logLineCallback(voidctx C.PVOID, str C.PCSTR) {
+	var ctx *logCallbackContext
+	ctx = (*logCallbackContext)(unsafe.Pointer(uintptr(voidctx)))
+	line := C.GoString(str)
+	if !strings.Contains(line, "\n") {
+		ctx.unfinished = ctx.unfinished + line
+		return
+	}
+	if len(ctx.unfinished) != 0 {
+		line = ctx.unfinished + line
+		ctx.unfinished = ""
+	}
+	lines := strings.Split(line, "\n")
+	start := int(0)
+	if !ctx.hasSeenRetAddr {
+		for idx, l := range lines {
+			if strings.HasPrefix(l, bugcheckCodePrefix) {
+				ctx.loglines = append(ctx.loglines, l)
+				return
+			}
+			if strings.HasPrefix(l, debugSessionTimePrefix) {
+				ctx.loglines = append(ctx.loglines, l)
+				return
+			}
+			if strings.HasPrefix(l, "RetAddr") {
+				ctx.hasSeenRetAddr = true
+				start = idx
+			}
+		}
+		if !ctx.hasSeenRetAddr {
+			return
+		}
+
+	}
+	ctx.loglines = append(ctx.loglines, lines[start:]...)
+}
+func parseCrashDump(wcs *WinCrashStatus) {
+	var ctx logCallbackContext
+	var extendedError uint32
+
+	err := C.readCrashDump(C.CString(wcs.FileName), unsafe.Pointer(&ctx), (*C.long)(unsafe.Pointer(&extendedError)))
+
+	if err != C.RCD_NONE {
+		wcs.Success = false
+		wcs.ErrString = fmt.Sprintf("Failed to load crash dump file %d %x", int(err), extendedError)
+		log.Errorf("Failed to open crash dump %s: %d %x", wcs.FileName, int(err), extendedError)
+		return
+	}
+
+	if len(ctx.loglines) < 2 {
+		wcs.ErrString = fmt.Sprintf("Invalid crash dump file %s", wcs.FileName)
+		wcs.Success = false
+		return
+	}
+
+	end := min(len(ctx.loglines)-1, maxscan)
+	for _, line := range ctx.loglines[:end] {
+		// skip lines that start with RetAddr, that's just the header
+		if strings.HasPrefix(line, debugSessionTimePrefix) {
+			parts := strings.SplitN(line, ":", 2)
+			if len(parts) == 2 {
+				wcs.DateString = strings.TrimSpace(parts[1])
+			}
+			continue
+		}
+
+		if strings.HasPrefix(line, bugcheckCodePrefix) {
+			codeAsString := strings.TrimSpace(line[len(bugcheckCodePrefix)+1:])
+			wcs.BugCheck = codeAsString
+			continue
+		}
+		if strings.HasPrefix(line, "RetAddr") {
+			continue
+		}
+		if strings.HasPrefix(line, "Unable to") { // "Unable to load image, which is ok
```

Comment:
I don't think this check is needed (at least, it isn't obvious to me why). If it is, could you explain? If it isn't, I think we can remove it.

Also, am I correct in assuming that this loop iterates the output (from your readme) and the things it cares to pull out include:
  - The time the debug session was kicked off
  - The bugcheck prefix
  - The first callsite that isn't `nt!`

?

Reply:
I added some sample lines in comment.  But there could be "unable to load image" interlaced with lines we want, in which case the search for `NT!` would fail but the line isn't the one we want.  Hopefully it's more clear with the sample lines comment to be added.